### ISSUE TITLE
ediff: request for "simple" ediff instead of 3 way merge for everything except merges

### ISSUE THREAD

**Issue by Silex** [magit/magit#1743]
Hello,

When doing `M-e` on staged or unstaged changes, magit currently shows a 3-way merge ediff.

A classic ediff like what `ediff-buffers` does is more sensible IMHO.

**Comment by jsyjr** [comment]
+1

**Comment by tarsius** [comment]
Does it really? What I see is that the user is presented with a default choice `A..B` (which she may change to `A...B` to explicitly compare with the merge base) and that `magit-ediff-compare--read-revisions` then returns a list of that range and the two revisions individually (which seems like an unnecessary complication, but that's not relevant here). Then two buffers are created showing the blobs from these two revisions, resulting in a regular "compare two buffers" ediff.

What am I missing? How do you get the 3-way ediff? I need exact steps - maybe under certain circumstances a bug in `magit-ediff-dwim` kicks in and `magit-ediff-resolve` or `magit-ediff-merge` is used instead of `magit-ediff-compare`. Also try whether the same happens if you use `magit-ediff-compare` (<kbd>M-d</kbd>) directly.

**Comment by jsyjr** [comment]
I have a bunch of changes that I want to review so as to understand what I have done:
* In the status dashboard I move my cursor to a file line
* Possibly I TAB to expand, possibly I cursor down to an ill-understood change
* I decide that I want to see my change(s) in context so I hit M-e (magit-ediff-dwim)

Currently magit opens ediff with three panes (A: ~HEAD~, B: workspace, C: ~[index]~) positioned at the top of file.  This is sub-optimal:
* Given magit's laudable effort to use stage / unstage terminology the ~[index]~ qualifier is confusing.  Pane labeling more consistent with the dashboard and magit conventions would be [HEAD], [unstaged] and [staged].  (Similarly since popup S is described as 'Stage all' popup U should be 'Unstage all', not 'Reset index'.)
* Positioning the unstaged changes pane in the center also conflicts with the dashboard.  I suggest left to right ordering: [unstaged], [staged], [HEAD].  (I could also live with right to left ordering.)
* Except when both both staged and unstaged changes are present only two panes should appear.  This way if magit actually opts to display three panes then one knows immediately that something atypically is afoot.
* If M-e happens with the cursor not on a file line but on an specific difference then ediff should open displaying that difference (not the top of file).

**Comment by tarsius** [comment]
@Silex Ah sorry, I somehow understood you were comparing commits and got three buffers.

After reading it again I now understand that you want two buffers when viewing *unstaged* changes, and two other buffers when viewing *staged* changes. But Magit in both cases intentionally shows the same three buffers.

I think that it makes sense to do it that way - and more on that below - but first the details raised by @jsyjr.

* If I remember correctly I named the buffers taking inspiration from VC, if not that then Emacs backup filenames. I intend to implement a file handler for Git blobs, until then that's an area I have opted to fill with a hack in order not to further delay the release.
* The order was indeed wrong, I have fixed that.
* See below.
* That would indeed be nice. But since we are in feature freeze it will have to wait until after the release. Please open a new issue for this.

So why are three buffers needed (it's not just because then I have only to implement one command instead of two)? It's critical that these buffers are not just for looking at `staged` and/or `unstaged` changes, but also to act on them, i.e. **stage**, **unstage** and **discard**. When you have modify the `staged` and/or `unstaged` buffer and quit the Ediff session you are asked whether you want to save them. For the `unstaged` buffer this just means saving to the worktree file (that buffer actually is just visiting that file) for the `staged` buffer, that means only now updating the index for that file with the buffer contents.

To view the *unstaged* changes and to **stage** them you need two buffers `staged` and `unstage`. Git, because it stores revisions not diffs, also needs to look at both stages to generate a diff.  "Diffs in Magit" doesn't need both stages because it tells Git to deal with it and then just displays the result, which isn't two revisions/buffers/sections but the diff. Ediff is not the same as diff, it shows the two stages and colors each one were it differs from the other. Since you use ediff, you know why that can be useful - in short: more context.

(In a way Ediff presents the intermediate result of what `diff`, `git diff` etc. do to generate the diff - it finds the areas that differ, but then instead of presenting just those, with little context; it presents both revisions in full, marking the areas that differ. This is more verbose which is why I generally prefer diffs over ediff, but sometimes you need the context. It also means, and that is quite relevant to this discussion, that to fully appreciate the "differences in one stage" the user needs to be able to see both stages. Otherwise you would get more context in the areas that don't matter so much - those that didn't change. But less context in those that matter most because the changed - without the other stage you would only know that something changed in an area - not how it changed.)

To view the *staged* changes and to **unstage** you need two buffers `HEAD` and `staged`. But  to **unstage (without risking losing work)** them you also need the `unstaged` buffer. In most cases - when the same area as not been further modified in the work tree - you don't need the third buffer. But to know whether that is the case you do need the third buffer, otherwise you might lose the intermediate, *staged* changes without realizing it.

When you realize, that no, what you just **unstaged** should have remained *staged*, then you also need the `unstaged` changes in addition to the `staged` ones, or you would not be able to **stage** them again, as in Ediff that is done by **copying** from `unstaged` to `staged`.

Also to *discard* `staged` changes (remove from index _and_ work tree) you need three buffers. That's not as convenient in Ediff as it is Magit because there is only **copy-A-to-B** and **copy-A-to-C**, but no (I believe) **copy-A-to-B-as-well-as-C**, similar to **magit-discard**.

----

In summary, `magit-ediff-stage` is about performing staging operations, which includes both staging and unstaging. Because Ediff doesn't present a "summary of the differences in two revisions" but "the two revisions, with the differences marked" to see one difference two buffers are needed. To enable this operations we need two comparisons. In order to present two comparisons we need three buffers (not four because the two comparisons share one of the buffers).

**Comment by Silex** [comment]
Hum, I always thought staging/unstaging should be done in the magit-status buffer, and that the ediff view was there as a convenience to view modifications side-by-side.

In my workflow I was using the ediff view to have a vertical side-by-side before/after view of the changes I did, reguardless of if it's staged/unstated/from a previous commit.

I understand that now ediff support is enhanced to allow all sort of operations (which is fine), like it does to help resolve merge.

So, basically to get the behavior I want I could simply remap `M-e` to `magit-ediff-compare` or whatever doesn't annoy me with the unnecessary views. I might event start to get used about the ediff view and use it for staging!

So for me this issue is almost closable as far as I'm concerned. I just didn't understand ediff was becoming an alternative for staging/unstaging.

**Comment by Silex** [comment]
Hum, I always thought staging/unstaging should be done in the magit-status buffer, and that the ediff view was there as a convenience to view modifications side-by-side.

In my workflow I was using the ediff view to have a vertical side-by-side before/after view of the changes I did, reguardless of if it's staged/unstated/from a previous commit.

I understand that now ediff support is enhanced to allow all sort of operations (which is fine), like it does to help resolve merge.

So, basically to get the behavior I want I could simply remap `M-e` to `magit-ediff-compare` or whatever doesn't annoy me with the unnecessary views. I might event start to get used about the ediff view and use it for staging!

So for me this issue is almost closable as far as I'm concerned. I just didn't understand ediff was becoming an alternative for staging/unstaging.

**Comment by Silex** [comment]
Hum, I always thought staging/unstaging should be done in the magit-status buffer, and that the ediff view was there as a convenience to view modifications side-by-side.

In my workflow I was using the ediff view to have a vertical side-by-side before/after view of the changes I did, reguardless of if it's staged/unstated/from a previous commit.

I understand that now ediff support is enhanced to allow all sort of operations (which is fine), like it does to help resolve merge.

So, basically to get the behavior I want I could simply remap `M-e` to `magit-ediff-compare` or whatever doesn't annoy me with the unnecessary views. I might event start to get used about the ediff view and use it for staging!

So for me this issue is almost closable as far as I'm concerned. I just didn't understand ediff was becoming an alternative for staging/unstaging.

**Comment by Silex** [comment]
Hum, I always thought staging/unstaging should be done in the magit-status buffer, and that the ediff view was there as a convenience to view modifications side-by-side.

In my workflow I was using the ediff view to have a vertical side-by-side before/after view of the changes I did, reguardless of if it's staged/unstated/from a previous commit.

I understand that now ediff support is enhanced to allow all sort of operations (which is fine), like it does to help resolve merge.

So, basically to get the behavior I want I could simply remap `M-e` to `magit-ediff-compare` or whatever doesn't annoy me with the unnecessary views. I might event start to get used about the ediff view and use it for staging!

So for me this issue is almost closable as far as I'm concerned. I just didn't understand ediff was becoming an alternative for staging/unstaging.

**Comment by tarsius** [comment]
> I just didn't understand ediff was becoming an alternative for staging/unstaging.

Yeah, I considered calling it a "documentation issue". That being said, I just implemented it, I don't use it (or anything Ediff really), so there may be bugs lingering. I think that it makes sense for Magit to only provide the do-staging Ediff, not also the looking-at-staged and looking-at-unstaged variants, because in Magit you can act on any diff, they are not just for looking at  - so the same should apply to its Ediff support.

(I do intend to use Ediff more in the future).

**Comment by tarsius** [comment]
Okay, I am closing this. Please try using the the three-buffer view for a while - don't just look at them, also experiment with staging. If after doing that for a while you still feel you want the just-looking-at variants, please say so. But in any case that won't be implemented before the release.

**Comment by jwiegley** [comment]
@tarsius How is one supposed to stage changes with `magit-ediff-stage`?  For me, the HEAD and index buffers are read-only, so the only operation I have available is to change the working tree copy to either the HEAD or the index version.

**Comment by tarsius** [comment]
Fixed in 95b02cc0cc64c542f0583e2d03902c2778ff2aba. Thanks for pointing it out.

**Comment by tarsius** [comment]
I have implemented to new ediff variants `magit-ediff-show-staged` and `magit-ediff-show-unstaged`. As the names imply these commands are only for looking at the differences. To stage, unstage, and discard using ediff use `magit-ediff-stage` which as I have explained above requires three buffers to be displayed, the new commands use only two.

No keys are bound to the new commands. To add them to the ediff popup use something like:

```lisp
(magit-define-popup-action 'magit-ediff-popup ?S "Show staged" 'magit-ediff-show-staged)
(magit-define-popup-action 'magit-ediff-popup ?U "Show unstaged" 'magit-ediff-show-unstaged)
```

`magit-ediff-dwim` never uses the new functions, as there is no way for it to tell whether you want to stage/unstage/discard using ediff, or just look at the differences using ediff.

**Comment by gseng** [comment]
@tarsius , I tried this on the latest Magit:

```
(use-package magit
  :demand
  :ensure magit
  :config
  (progn

    ;;https://github.com/magit/magit/issues/1743                                                                                                                                                               
    (magit-define-popup-action 'magit-ediff-popup ?S "Show staged" 'magit-ediff-show-staged)
    (magit-define-popup-action 'magit-ediff-popup ?U "Show unstaged" 'magit-ediff-show-unstaged)
    )
)
(provide 'my-magit)    
```

But got this:
```
Error (use-package): magit :config: Symbol's value as variable is void: magit-ediff-popup
```

Any ideas? Where should I put those lines?

Thanks.

**Comment by npostavs** [comment]
Your `use-package` form worked for me.

> Error (use-package): magit :config: Symbol's value as variable is void: magit-ediff-popup

Maybe you loaded an old version of magit which doesn't define `magit-ediff-popup`? Check your `load-path`.

**Comment by gseng** [comment]
@npostavs , I do have the latest version. Just curious, did you try it with a "fresh" instance of emacs?

I think it has something to do with getting [`magit-ediff-popup`](https://github.com/magit/magit/blob/master/lisp/magit-ediff.el#L59) to be in "scope" or loaded ... unfortunately I don't understand emacs/elisp enough yet to be able to figure out how do accomplish this.

This is what I see:
* Add `use-package` config
* Open fresh emacs
* See error
* Invoke `magit-status`
* Re-run `use-package`
* See error
* In `magit-status` buffer, press `E` to get ediff popup
* Re-run `use-package`
* Works now (`?S` and `?U` are added to popup)

Thanks.

**Comment by tarsius** [comment]
Requiring `magit` does not load `magit-ediff`. If you use Elpa, then the commands from `magit-ediff` should get autoloaded though. So you should be able to just use them without explicitly requiring `magit-ediff`. Otherwise you might have to explicitly load either `magit-ediff` (`(use-package magit-ediff ...)`) or `magit-autoloads.el` (which is generated by `make lisp`).

**Comment by npostavs** [comment]
> Requiring magit does not load magit-ediff

Hmm mine does. Oh I see, I didn't build magit-autoloads in my src testing copy:
```elisp
;; near bottom of magit.el
  (unless (load "magit-autoloads" t t)
    (require 'magit-ediff)
```

**Comment by jsyjr** [comment]
I load magit via el-get.  Here is the recipe:

(:name magit
>        :website "https://github.com/magit/magit#readme"
>        :description "It's Magit! An Emacs mode for Git."
>        :type github
>        :pkgname "magit/magit"
>        :branch "master"
>        :minimum-emacs-version "24.4"
>        :depends (dash)
>        :provide (with-editor)
>        :info "Documentation"
>        :load-path "lisp/"
>        :compile "lisp/"
>        ;; use the Makefile to produce the info manual, el-get can
>        ;; handle compilation and autoloads on its own.
>        :build `(("make" ,(format "EMACSBIN=%s" el-get-emacs) "docs"))
>        :build/berkeley-unix `(("gmake" ,(format "EMACSBIN=%s"
> el-get-emacs) "docs"))
>        ;; assume windows lacks make and makeinfo
>        :build/windows-nt (progn nil))


Out of the box it compiles magit-ediff.el.  In my .emacs I have:

(add-hook 'magit-mode-hook
>           (lambda ()
>             (magit-define-popup-action 'magit-ediff-popup ?S "Show staged"
>   'magit-ediff-show-staged)
>             (magit-define-popup-action 'magit-ediff-popup ?U "Show
> unstaged" 'magit-ediff-show-unstaged)))


Works just fine.

/john

**Comment by gseng** [comment]
Thanks all for your help. This worked for me:

```
(use-package magit
  :ensure magit
  :config
  (progn
    (require 'magit-ediff)
    ;;https://github.com/magit/magit/issues/1743                                                                                                                                                               
    (magit-define-popup-action 'magit-ediff-popup ?S "Show staged" 'magit-ediff-show-staged)
    (magit-define-popup-action 'magit-ediff-popup ?U "Show unstaged" 'magit-ediff-show-unstaged)
    )
)
(provide 'my-magit)    
```

@jsyjr , `el-get` looks great, might port my emacs inits over to that someday.

**Comment by npostavs** [comment]
> I load magit via el-get. Here is the recipe:

Hmm, actually this might not be making `magit-autoloads` either, I'll see if I can fix that.

**Comment by npostavs** [comment]
> If you use Elpa, then the commands from magit-ediff should get autoloaded though. So you should be able to just use them without explicitly requiring magit-ediff

@tarsius: the command `magit-ediff-popup` is autoloaded, but the variable is not, which is why `magit-define-popup-action` fails. Should we change `magit-define-popup-action` to store pending definitions if the variable hasn't been loaded yet?

Barring that, using
```elisp
(with-eval-after-load 'magit-ediff
  (magit-define-popup-action 'magit-ediff-popup ?S "Show staged" 'magit-ediff-show-staged)
  (magit-define-popup-action 'magit-ediff-popup ?U "Show unstaged" 'magit-ediff-show-unstaged))
```
should work, or in use-package terms:
```elisp
(use-package magit
  ...
  :config
  (use-package magit-diff
    :defer t
    :config
    (magit-define-popup-action 'magit-ediff-popup ?S "Show staged" 'magit-ediff-show-staged)
    (magit-define-popup-action 'magit-ediff-popup ?U "Show unstaged" 'magit-ediff-show-unstaged)))
```

@jsyjr: you will have the same problem as gseng after dimitri/el-get#2222 is merged.

**Comment by jsyjr** [comment]
Thanks for the heads-up.

/john

**Comment by tarsius** [comment]
> Should we change magit-define-popup-action to store pending definitions if the variable hasn't been loaded yet?

Yes, I think that would be a good idea.

**Comment by liujoey** [comment]
My humble opinion, excuse my language, English is not my first language.

I use ediff to view the side-by-side diffs a lot. The current three panels implementation is really confusing. 

First of all, the ~HEAD~ and ~{index}~ tag are put after the file name, when you have a deep path it makes it harder to know which panel is which. Why don't we put these tags before the file name?

Second, in the magit status window, when I put my cursor on a file in unstaged session and execute magit-ediff-show-unstaged, it should automatically pick the file under cursor instead of ask me to type the file name in, shouldn't it?

Third, I really would love to press one same key, `I` maybe, when cursor on a unstaged file it run `magit-ediff-show-unstage` and when the cursor on a staged file it run `magit-ediff-show-stage`.

Last. If a file is unstaged only, open ediff on it should only show two panel which ~{index}~ on left and working tree on right. If a file is staged only, open ediff on it should only show two panel which ~HEAD~ on left and ~{index}~ on right, you don't need the unstaged panel because it's content is same as ~{index}~. Only if I'm doing conflicts solving or I have both staged and unstaged hunks on same file, I would need the 3 ways diffs.

The diff view in vim-fugitive makes a lot sense.

**Comment by stefanq** [comment]
A few weeks ago just pressing 'e' on staded or unstaged changes invoked a 'normal' two buffer ediff.
1) No longer. For some reason magit (melpa 20160509.1609) currently will show a 3-way merge ediff, which makes it impossible to use ediff functionality to modify either buffer A or B.

2) Pressing 'E' gives me a magit-ediff-popup buffer, but it does not contain any ediff options apart from 'Dwim', which invokes the same inappropriate three-way-diff.

According to the magit major mode help (Ctrl-h-m) apart from 'e' and 'E' magit offers no other ediff keybinding. Which would be Ok if the magit-ediff-popup would provide additional magit-ediff commands which obviously do exist (but are rather disconnected from the magit UI).

I'd consider the current state of ediff integration is of really bad usability.

**Comment by tarsius** [comment]
> 1

You need to be more explicit. Did it previously use a 2-buffer diff? What's the value of `magit-ediff-dwim-show-on-hunks`? (I admit the name of that option isn't very descriptive.) Are the buffers really read-only?

> 2

`E` should show a popup with all the commands mentioned in the manual. I cannot see a reason why that could possibly not be the case.

> I'd consider the current state of ediff integration is of really bad usability.

The behavior you are seeing is not the expected behavior. Something is wrong, and I suspect it is your configuration (or possibly the "starter-kit" you are using). [Please try to reproduce the issue using `emacs -Q`](http://www.lunaryorn.com/2015/11/29/reproduce-bugs-in-emacs-Q.html).

**Comment by stefanq** [comment]
> You need to be more explicit. Did it previously use a 2-buffer diff?

Yes just pressing `e` on a changed file did this. But no longer after a downgrade (reinstallation) from openSUSE Tumbleweed to Leap or a later emacs update (all i know is that I'm currently on 24.5-233.5).

> What's the value of magit-ediff-dwim-show-on-hunks?

I first found this variable not known to my emacs. Obviously this and others come into existence after doing the first magit-ediff command. After that its value is `nil`.
Setting it results in `e` triggering a two-way diff.
May be I can make this persist with an addtion to the init file as proposed in other comments.

> Are the buffers really read-only?

Only the 'HEAD' file variant, but this three-way ediff (with all three buffers vertically) just hasn't the usual commands of the two-way ediff for copying a current diff from A to B (or vice versa)).

> `E` should show a popup with all the commands mentioned in the manual.

I now found that the several Actions named "Show ..."  in this menu are actually doing the two-way ediffs which I was looking for.

> The behavior you are seeing is not the expected behavior.

I'm glad to here this.

> Please try to reproduce the issue using `emacs -Q`.

Did this, followed by `M-x package-initialize` and `magit-status`: After pressing `e` on any modified  file I get the same three way diff of a HEAD, index and the regular file variant (regardless of wether the file is staged or unstaged).

**Comment by tarsius** [comment]
> May be I can make this persist with an addtion to the init file as proposed in other comments.

Yes, that's what you should do.

**Comment by liujoey** [comment]
yeah, setting `magit-ediff-dwim-show-on-hunks` to `t` did the trick.
